### PR TITLE
New function for Jacobi polynomial evaluation and its roots

### PR DESCRIPTION
**Pull request by kronbichler**
This PR makes the computation of Jacobi polynomials, that was previously inside the computation of Gauss-Lobatto quadrature points, externally visible. This has various benefits:
1. With the update I can now also use the Jacobi root function for the Gaussian quadrature. Due to an improvement on how the recursion inside Jacobi polynomials is done, the new variant is now almost as fast as the old hand-coded version in the Gaussian quadrature, but it saves 50 lines of code.

2. The roots of the Jacobi polynomial with `alpha=beta=2` are also used for the HermiteLikeInterpolation introduced in #5879. This is cheaper than the LAPACK variant we had before, slightly more accurate (15 digits correct rather than 12-13) and avoids that dependency in that case.

3. One could, in principle, use the functions also with arbitrary precision, not just `long double` or `double` as we currently do. For example, one could use `float128` if for some reason there is a need for getting even more digits...

### CODE REVIEW

**Comment by kronbichler** [comment]
Any more comments to this PR?

File: source/base/polynomial.cc [hunk 1]
```diff
@@ -1395,34 +1393,13 @@ namespace Polynomials
         // entries given below. For degree 4 the eigenvalue is zero, so bypass
         // the LAPACK logic in that case.
```
Comment (drwells) [inline source/base/polynomial.cc]: I think this comment needs to be updated since the function no longer depends on LAPACK.

**Comment by kronbichler** [comment]
> Would it be possible to change `jacobi_polynomial_value` and `jacobi_polynomial_roots` to use `[0, 1]` instead of `[-1,1]`?

I could in principle adjust them, yes, by scaling the input in `jacobi_polynomial_value` back to `[-1, 1]` and adjusting the root finding. We need to be careful when translating between the two intervals because the recursion is in `[-1, 1]`. I will try that tomorrow.

**Comment by kronbichler** [comment]
@drwells It is indeed much nicer with `[0, 1]` at the places where we use the results of these polynomials/roots. I have changed the polynomials and their use.

File: include/deal.II/base/polynomial.h [hunk 2]
```diff
@@ -709,6 +707,43 @@ namespace Polynomials
...
   };
+
+
+
+  /*
+   * Evaluate a Jacobi polynomial $ P_n^{\alpha, \beta}(x) $ specified by the
+   * parameters @p alpha, @p beta, @p n, where @p n is the degree of the
+   * Jacobi polynomial.
+   *
+   * @note The Jacobi polynomials are not orthonormal and defined on the unit
```
Comment (drwells) [inline include/deal.II/base/polynomial.h]: should be 'and are defined'

File: source/base/quadrature_lib.cc [hunk 3]
```diff
@@ -340,16 +145,17 @@ QGaussLobatto<1>::QGaussLobatto (const unsigned int n)
...
-    = internal::QGaussLobatto::compute_quadrature_points(n, 1, 1);
+    = Polynomials::jacobi_polynomial_roots<long double>(n-2, 1, 1);
+  points.insert(points.begin(), 0);
+  points.push_back(1.);
   std::vector<long double> w
     = internal::QGaussLobatto::compute_quadrature_weights(points, 0, 0);
 
-  // scale points to the interval
-  // [0.0, 1.0]:
+  // scale points to the interval [0.0, 1.0]:
```
Comment (drwells) [inline source/base/quadrature_lib.cc]: technicall this should just say 'scale weights'

File: tests/base/polynomial_jacobi_roots.cc [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                  << std::setw(9) << std::setprecision(3)
+                  << jacobi_polynomial_value(degree, alpha, beta, roots[i]) << " ) "
+                  << std::setw(22) << std::setprecision(16)
+                  << 0.5 + 0.5 * roots_reference[i] << " (fval = "
+                  << std::setw(9) << std::setprecision(3)
+                  << jacobi_polynomial_value(degree, alpha, beta,
+                                             0.5 + 0.5 * roots_reference[i])
+                  << " )" << std::endl;
+        deallog << std::endl;
+      }
```
Comment (drwells) [inline tests/base/polynomial_jacobi_roots.cc]: This is a very nice test.

**Comment by drwells** [comment]
32c489a90201a7bfe657dcaf5a1d11473e9472cf in particular is quite nice: its always good to see these calculations simplified and generalized.

**Comment by kronbichler** [comment]
Thanks for the comments @drwells. It should be fixed now.

**Comment by drwells** [comment]
Thanks.

Nobody else has commented so I'll go ahead and mark this as ready to merge.

**Comment by kronbichler** [comment]
> Still some test failures:

That's funny because what I did was to increase the precision with the suggestion by @drwells: The reason is that we previously computed the Gaussian quadrature points (and Gauss-Lobatto support points) as `0.5 + 0.5*static_cast<double>(jacobi_points[i])` in `double` precision whereas they are now (intrinsically) done in `long double` precision. I'll have to re-run the whole suite on my system to see if there are more failures now.

**Comment by kronbichler** [comment]
On the other hand, this is a well-tested commit :smile:
Getting the Jacobi polynomials or their zeros wrong now means that the Gaussian quadrature is wrong, and that leads to lots of test failures I can assure. :laughing:

**Comment by drwells** [comment]
I reran the full test suite yesterday and did not notice any test failures. I am surprised that changing the reference interval changed this much.

**Comment by kronbichler** [comment]
I ran all tests on my machine and they are passing now.

**Comment by davydden** [comment]
> Still some test failures:

I also noticed some difference in my own calculations ( FE_Q(2) + matrix-free + GMG converged with slightly different number of steps), bisected to this commit 6af60d8a90bb595aed7f601e814019361f49a774 . 

This is probably due to

> The reason is that we previously computed the Gaussian quadrature points (and Gauss-Lobatto support points) as 0.5 + 0.5*static_cast<double>(jacobi_points[i]) in double precision whereas they are now (intrinsically) done in long double precision.

**Pull request merged**